### ISSUE TITLE
前端学习AI应用开发路线

### ISSUE THREAD

**Issue by giscafer** [giscafer/blog#72]
作为前端转型 AI 应用开发，你需要补充以下核心知识体系，并逐步建立 AI 工程化思维。以下是系统化学习路径和资源推荐：

---

#### 1. **机器学习基础**
+ **必学概念**：
    - 监督学习 vs 无监督学习
    - 神经网络基础（感知机、激活函数、反向传播）
    - 损失函数与优化器（梯度下降、Adam）
    - 过拟合与正则化（Dropout、L1/L2）
+ **推荐资源**：
    - 📚 书籍：[《机器学习实战：基于 Scikit-Learn、Keras 和 TensorFlow》](https://book.douban.com/subject/35260563/)（实践导向）
    - 🎥 课程：[吴恩达《机器学习》](https://www.coursera.org/learn/machine-learning)（经典理论基础）
    - 📄 图解教程：[《Visualizing Machine Learning Concepts》](https://jalammar.github.io/)（适合视觉化学习）
    - [aka.ms/ai-agents-beginners](https://aka.ms/ai-agents-beginners)
    - [microsoft/generative-ai-for-beginners](https://github.com/microsoft/generative-ai-for-beginners?WT.mc_id=academic-105485-koreyst)
    - [赋范2025大模型技术全景地图](https://kq4b3vgg5b.feishu.cn/wiki/MTuywZLzhioQhHk2yolcjgawnTg)

#### 2. **深度学习核心**
+ **关键领域**：
    - NLP（Transformer、BERT、GPT 架构）
    - 计算机视觉（CNN、目标检测、图像生成）
    - 生成式模型（Diffusion Models、GAN）
+ **学习工具**：
    - 🛠️ [TensorFlow Playground](https://playground.tensorflow.org/)（神经网络可视化实验）
    - 🧩 [Colab Notebooks](https://colab.research.google.com/)（快速运行 AI 代码）

---

#### 1. **AI 应用架构设计**
+ **典型架构模式**：

```bash
前端 → API 网关 → 模型服务 → 数据存储
            ↘ 缓存层 ↗
```

+ **关键技术点**：
    - 模型服务化（REST/gRPC 接口）
    - 流式传输（SSE/WebSocket）
    - 边缘计算（WebAssembly + ONNX 模型）

#### 2. **大模型应用开发**
+ **核心能力**：
    - Prompt Engineering（结构化提示词设计）
    - RAG（检索增强生成）架构实现
    - Fine-tuning（基于 LoRA 的高效微调）
+ **实践工具**：
    - 🔧 [LangChain](https://python.langchain.com/)（AI 应用编排框架）
    - 🧠 [LlamaIndex](https://www.llamaindex.ai/)（知识库增强检索）
    - ⚡ [OpenAI API](https://platform.openai.com/docs)（商业模型快速接入）
    - [Model Context Protocol (MCP) ](https://modelcontextprotocol.io/introduction)

#### 3. **浏览器端 AI**
+ **关键技术**：
    - TensorFlow.js 模型部署
    - WebGL 加速推理
    - ONNX Runtime Web 集成
+ **实战项目**：
    - 实现浏览器内图像分类（MobileNet）
    - 文字生成（GPT-2 量化模型）
    - 示例代码：

```javascript
import * as tf from '@tensorflow/tfjs';
const model = await tf.loadLayersModel('model.json');
const pred = model.predict(tensorInput);
```

---

#### 1. **AI 开发工具**
| 工具类型 | 推荐选择 | 学习重点 |
| --- | --- | --- |
| **开发框架** | PyTorch / TensorFlow.js | 模型部署与转换 |
| **模型仓库** | Hugging Face / Model Zoo | 预训练模型下载与微调 |
| **可视化工具** | Weights & Biases / TensorBoard | 训练过程监控 |
| **边缘计算** | ONNX Runtime / WebDNN | 浏览器端推理优化 |

#### 2. **全栈技术延伸**
+ **后端技能**：
    - Python FastAPI（模型服务化）
    - Node.js + TensorFlow.js（全 JS 栈 AI）
    - GraphQL（复杂数据查询）
+ **DevOps**：
    - 模型容器化（Docker + Triton Inference Server）
    - 监控（Prometheus + Grafana）

---

#### 1. **分阶段学习计划**

![举例阶段划分](https://github.com/user-attachments/assets/0fbfa73d-b998-4942-86b1-0906ee974673)

#### 2. **精选资源列表**
+ **免费课程**：
    - [Fast.ai《面向程序员的深度学习》](https://course.fast.ai/)（实战优先）
    - [Google《生成式 AI 学习路径》](https://cloud.google.com/learn/pathways/generative-ai)（大模型专项）
+ **付费课程**：
    - [Udacity AI 纳米学位](https://www.udacity.com/school-of-ai)（系统化项目实战）
+ **开发文档**：
    - [Hugging Face 文档](https://huggingface.co/docs)
    - [TensorFlow.js 指南](https://www.tensorflow.org/js/guide)
+ **社区与资讯**：
    - [Papers With Code](https://paperswithcode.com/)（最新论文+代码）
    - [AI Weekly](https://aiweekly.co/)（行业动态简报）

---

#### 1. **入门级项目**
+ **智能聊天助手**：
    - 技术栈：Next.js + OpenAI API + LangChain
    - 扩展点：增加知识库检索（LlamaIndex）
+ **浏览器图像处理**：
    - 技术栈：TensorFlow.js + MobileNet
    - 功能：实时滤镜风格迁移

#### 2. **进阶项目**
+ **文档智能分析工具**：
    - 技术栈：Python + LangChain + Pinecone
    - 功能：PDF 解析与语义搜索
+ **低代码 AI 工作流平台**：
    - 技术栈：React + FastAPI + Hugging Face
    - 功能：拖拽式模型编排

---

### **六、避坑指南**
1. **不要过度深入数学**：  
聚焦应用层开发时，理解算法输入输出和调参方法比推导公式更重要。
2. **警惕技术镀金**：  
优先掌握主流框架（PyTorch/TensorFlow），而非追逐最新论文模型。
3. **重视数据工程**：  
实际项目中 60% 时间会花在数据清洗和特征工程上。
4. **工程化思维**：  
学会设计可扩展的 AI 系统架构，而非只关注模型准确率。

---

转型 AI 应用开发的关键在于 **将前端工程化能力与 AI 模型落地经验结合**。保持对新技术的好奇心，同时深耕 1-2 个垂直领域（如 AIGC 工具开发、智能客服系统），逐步建立技术壁垒。
